zad2/warriors.rb

```diff
@@ -1,12 +1,114 @@
-class Warrior
+require 'faker'
+require 'ordinalize'
+
+# Base class for characters
+class Character
+  attr_reader :name
+  attr_accessor :level
+
   def initialize(name:, level:)
+    @name = name
+    @level = level
+    validate!
+  end
+
+  def strength(dice_roll)
+    @level + dice_roll
+  end
+
+  def card
+    "#{@name} (lvl #{@level})"
   end
 
-  def strength
+  def roll_1k12
+    rand(1..12)
+  end
+
+  private
+
+  def validate!
+    raise ArgumentError, 'Level should be a number between 1 and 99' unless (1..99).cover?(level.to_i)
   end
 end
 
+# Warrior character class
+class Warrior < Character
+  def level_up!(opponent_level)
+    @level += (opponent_level - @level + 1)
+    puts "#{name} has reached level #{level}!"
+  end
+end
+
+# Monster character class
+class Monster < Character; end
+
+# Battle Arena for characters to fight on
 class BattleArena
   def initialize(first_character, second_character)
+    @first_character = first_character
+    @second_character = second_character
+  end
+
+  def battle!
+    strength = set_strength
+
+    puts "#{@first_character.name} attacked #{@second_character.name} with #{strength[:first]} strength"
+    puts "#{@second_character.name} attacked #{@first_character.name} with #{strength[:second]} strength"
+
+    winner, loser = results(strength)
+
+    winner ? puts("The winner is #{winner.name}!") : puts('Tie!')
+
+    winner.level_up!(loser.level) if winner.class == Warrior &&
+                                     winner.level <= loser.level
+  end
+
+  private
+
+  def set_strength
+    { first: @first_character.strength(@first_character.roll_1k12),
+      second: @second_character.strength(@second_character.roll_1k12) }
+  end
+
+  def results(strength)
+    case strength[:first] <=> strength[:second]
+    when 1 then [@first_character, @second_character]
+    when -1 then [@second_character, @first_character]
+    end
   end
 end
+
+# Game engine
+class Game
+  @@battle_no = 1
```

Comment:
zauważ, że to jest zmienna w obrębie całej klasy, a nie instancji, czyli za każdym kiedy wywołamy `Game.new.battle!` battle_no będzie się podnosił. Zmiennych tego typu rubiowcy raczej nie darzą szczególną sympatią.

czy nie lepiej gdyby `game1 = Game.new` i `game2 = Game.new` miały osobne liczniki?